rfd/xxxx-bot-for-cert-renewals.md

```diff
@@ -0,0 +1,633 @@
+
+## What
+
+RFD TBD defines the high-level goals and architecture for a new agent that can
+continuously renew Teleport-issued certificates.
+
+## Why
+
+Teleport nodes running the SSH service maintain a connection to the cluster and
+are able automatically renew their certificates when a CA rotation takes
+place or the certificate is approaching its expiration date.
+
+Unfortunately, external systems that join a Teleport cluster (OpenSSH servers,
+databases, applications, etc.) are not under our control, and therefore do not
+have a built in mechanism to renew certificates under these circumstances. The
+workflow for working with these types of external systems is to generate
+certificates via `tctl auth sign`, copy them to the target, and manually
+configure the target system to use these certificates. This is a process that
+must be repeated whenever a certificate is about to expire. Additionally, this
+approach is likely to incur downtime during
+[CA rotations](https://goteleport.com/docs/architecture/authentication/#certificate-rotation)
+as the renewal has to be performed within the rotation's grace period in order
+to maintain connectivity.
+
+This experience can incentivize users to make decisions that conflict with
+security best practices, including:
+
+- Requesting certificates with a very large time-to-live (TTL)
+- Electing not to perform CA rotations
+
+Both of these decisions are simply delaying the inevitable - eventually, the
+certificates will need to be renewed and external systems will need to be
+updated. The agent proposed in this RFD intends to automate the process of
+renewing these certificates - requiring less manual work on behalf of our users,
+and enabling them to use shorter TTLs and perform frequent CA rotations.
+
+## Details
+
+### Goals
+
+The primary user this agent is designed for is the admin who wants to leverage
+Teleport for as many systems as possible, but simply does not have the time to
+keep track of all certificates and manually renew them.
+
+While users with sophisticated automation and dedicated credential management
+tools may benefit from this solution, they likely already have the resources to 
+automate certificate renewals.
+
+As a result, a primary goal for this tool is simplicity. We will optimize the user
+experience for those looking to add systems to their cluster quickly and easily rather
+than supporting a multitude of configuration options and hooks for customization.
+
+### Overview
+
+We will develop a new lightweight tool, `tbot`, which runs as an agent on
+external systems that are part of a Teleport cluster (a VM running a Postgres
+database, an OpenSSH server, or a CI/CD worker, for example).
+
+This agent will join the Teleport cluster, request certificates, store them in a
+user-specified destination, and then monitor the certificates and renew them as
+necessary.
+
+#### User Experience
+
+First, register a bot with the new `tctl bots add` subcommand, giving the bot a
```

Comment:
The enrollment endpoint to add bots (through `tctl bots add`) using invite codes should be protected by appropriate rate-limiting mechanisms, just like the other token-based invitations.

Reply:
I've added an appendix detailing the new gRPC functions to be added and noted the need for ratelimiting on `GenerateInitialRenewableUserCerts()`, which is where the admin-created token is exchanged for renewable certificates.

Reply:
Small change here, we've removed the `GenerateInitialRenewableUserCerts()` endpoint entirely in favor of adding support for what we need in Teleport's standard `auth.Register()` flow.